### PR TITLE
feat(config): replace secrets in global config

### LINKED ISSUE #12144: Secrets in the globalConfig (e.g. customEnvVariables) are not replaced
### How are you running Renovate?

Self-hosted

### Please select which platform you are using if self-hosting.

GitHub Enterprise Server

### If you're self-hosting Renovate, tell us what version of Renovate you run.

28.0.3

### Describe the bug

Assuming a bot config like this:
```
{
  binarySource: 'docker',
  customEnvVariables: {
    NPM_TOKEN: '{{ secrets.NPM_TOKEN }}',
  },
  secrets: {
    NPM_TOKEN: process.env.NPM_TOKEN,
  }
}
```
The use case for this would be to inject custom secrets into the sidecar containers that Renovate creates.
Adding the secret in the `secrets` field prevents Renovate from leaking the credential in the log when it outputs the config for a run or a repo.

Unfortunately Renovate will extract the global config from the given config before replacing the secrets.
This results in a secret replaced repo config, but in an unresolved global config.
So Renovate will not inject the secret into the container but literally `{{ secrets.NPM_TOKEN }}`.

Ideally we would like to have both configs replaced.

I attempted to fix it in #12138 but as @viceice correctly mentioned this should be tested with a real repo.

By looking through the code I saw that with my fix repo configs will not be replaced, because they are added after the secrets will be replaced.

I would suggest to add the secret replacement twice; Once when the global config is loaded and once when the repo config is loaded.

### Relevant debug logs

_No response_

### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by Chumper**
Reapplies #12138 with added test cases and additional parameters.
Default behaviour is the same, secrets getting deleted.
New behaviour is that secrets are applied without deleting them globally

## Changes:

Applies secret in global config.

## Context:

#12387
#12389
#12138
Closes #12144

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [X] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/secrets.spec.ts [hunk 1]
```diff
@@ -97,5 +97,25 @@ describe('config/secrets', () => {
...
     });
+    it('replaces secrets in a array of objects without deleting them', () => {
+      const config = {
+        secrets: { ARTIFACTORY_TOKEN: '123test==' },
+        hostRules: [
+          { hostType: 'npm', token: '{{ secrets.ARTIFACTORY_TOKEN }}' },
+        ],
+      };
+      const res = applySecretsToConfig(config, undefined, false);
+      expect(res).toMatchSnapshot();
```
Comment (viceice) [inline lib/config/secrets.spec.ts:108]: Please use toEqual matcher, as those objects are very simple.

File: lib/config/secrets.spec.ts [hunk 2]
```diff
@@ -97,5 +97,25 @@ describe('config/secrets', () => {
...
+      expect(Object.keys(res)).toContain('secrets');
+    });
+    it('replaces secrets in a array of strings without deleting them', () => {
+      const config = {
+        secrets: { SECRET_MANAGER: 'npm' },
+        allowedManagers: ['{{ secrets.SECRET_MANAGER }}'],
+      };
+      const res = applySecretsToConfig(config, undefined, false);
+      expect(res).toMatchSnapshot();
```
Comment (viceice) [inline lib/config/secrets.spec.ts:117]: Same here.

File: lib/config/secrets.ts [hunk 3]
```diff
@@ -115,13 +122,14 @@ function replaceSecretsinObject(
 
 export function applySecretsToConfig(
   config: RenovateConfig,
-  secrets = config.secrets
+  secrets = config.secrets,
```
Comment (viceice) [inline lib/config/secrets.ts:125]: Do we use this parameter at any calling function? Or do we always use this default value?

Comment (Chumper) [reply lib/config/secrets.ts:125]: We use it here: https://github.com/renovatebot/renovate/blob/main/lib/workers/repository/init/merge.ts#L232

File: lib/config/secrets.spec.ts (hunk 2 above)
Comment (Chumper) [reply lib/config/secrets.spec.ts]: Done

File: lib/config/secrets.spec.ts (hunk 1 above)
Comment (Chumper) [reply lib/config/secrets.spec.ts]: Done

File: lib/workers/repository/index.ts [hunk 4]
```diff
@@ -23,7 +24,9 @@ export async function renovateRepository(
   canRetry = true
 ): Promise<ProcessResult> {
   splitInit();
-  let config = GlobalConfig.set(repoConfig);
+  let config = GlobalConfig.set(
+    applySecretsToConfig(repoConfig, undefined, false)
```
Comment (viceice) [inline lib/workers/repository/index.ts:28]:
```suggestion
    applySecretsToConfig(repoConfig, {}, false)
```
I think we need to pass an empty object to be safe for strict null checks

File: lib/config/secrets.ts [hunk 5]
```diff
@@ -85,21 +85,28 @@ function replaceSecretsInString(
 
 function replaceSecretsinObject(
   config_: RenovateConfig,
-  secrets: Record<string, string> = {}
+  secrets: Record<string, string> = {},
+  deleteSecrets = true
```
Comment (viceice) [inline lib/config/secrets.ts:89]:
```suggestion
  secrets: Record<string, string>,
  deleteSecrets
```

I think they are always passed?

Comment (Chumper) [reply lib/config/secrets.ts:89]: They are not, we have occurrences where we just pass the config and nothing else.
AFAIK for TS there is no way to use named parameters, so all optional parameters are required if we want to pass `deleteSecrets`
However, I can change all calls to use all parameters.

File: lib/config/secrets.ts [hunk 6]
```diff
@@ -85,21 +85,28 @@ function replaceSecretsInString(
 
 function replaceSecretsinObject(
```
Comment (viceice) [inline lib/config/secrets.ts:86]: @zharinov We should use  uppercase for `i` in a follow up PR. 😉
```suggestion
 function replaceSecretsInObject(
```

**Review by viceice — APPROVED** [review]

File: lib/config/secrets.ts (hunk 6 above)
Comment (Chumper) [reply lib/config/secrets.ts:86]: https://github.com/renovatebot/renovate/issues/13861

**Pull request merged**